src/Kentico.Xperience.Algolia/Indexing/DefaultAlgoliaTaskProcessor.cs

```diff
@@ -0,0 +1,124 @@
+        private readonly IEventLogService eventLogService;
+        private readonly IWebPageUrlRetriever urlRetriever;
+
+        public DefaultAlgoliaTaskProcessor(IAlgoliaClient algoliaClient,
+            IEventLogService eventLogService,
+            IWebPageUrlRetriever urlRetriever,
+            IServiceProvider serviceProvider)
+        {
+            this.algoliaClient = algoliaClient;
+            this.eventLogService = eventLogService;
+            this.serviceProvider = serviceProvider;
+            this.urlRetriever = urlRetriever;
+        }
+
+        /// <inheritdoc />
+        public async Task<int> ProcessAlgoliaTasks(IEnumerable<AlgoliaQueueItem> queueItems, CancellationToken cancellationToken)
+        {
+            var successfulOperations = 0;
+
+            // Group queue items based on index name
+            var groups = queueItems.GroupBy(item => item.IndexName);
+            foreach (var group in groups)
+            {
+                try
+                {
+                    var deleteIds = new List<string>();
+                    var deleteTasks = group.Where(queueItem => queueItem.TaskType == AlgoliaTaskType.DELETE).ToList();
+
+                    var updateTasks = group.Where(queueItem => queueItem.TaskType is AlgoliaTaskType.PUBLISH_INDEX or AlgoliaTaskType.UPDATE);
+                    var upsertData = new List<JObject>();
+                    foreach (var queueItem in updateTasks)
+                    {
+                        var documents = await GetDocument(queueItem);
+                        if (documents is not null)
+                        {
+                            foreach (var document in documents)
+                            {
+                                upsertData.Add(document);
+                            }
+                        }
+                        else
+                        {
+                            deleteTasks.Add(queueItem);
+                        }
+                    }
+                    deleteIds.AddRange(GetIdsToDelete(deleteTasks ?? new List<AlgoliaQueueItem>()).Where(x => x is not null).Select(x => x ?? ""));
+
+                    successfulOperations += await algoliaClient.DeleteRecords(deleteIds, group.Key, cancellationToken);
+                    successfulOperations += await algoliaClient.UpsertRecords(upsertData, group.Key, cancellationToken);
+                }
+                catch (Exception ex)
+                {
+                    eventLogService.LogError(nameof(DefaultAlgoliaClient), nameof(ProcessAlgoliaTasks), ex.Message);
+                }
+            }
+
+            return successfulOperations;
+        }
+
+        /// <inheritdoc/>
+        public async Task<IEnumerable<JObject>?> GetDocument(AlgoliaQueueItem queueItem)
+        {
+            var algoliaIndex = AlgoliaIndexStore.Instance.GetIndex(queueItem.IndexName) ?? throw new Exception($"AlgoliaIndex {queueItem.IndexName} not found!");
```

Comment:
Use `AlgoliaIndexStore.GetRequiredIndex()` which throws the exception for you.

Reply:
Changed